File: codes/leetcode/AllPathsFromSourcetoTarget.cpp
/*
************************************************
username    :   smmehrab
fullname    :   s.m.mehrabul islam
email       :   mehrab.24csedu.001@gmail.com
institute   :   university of dhaka, bangladesh
session     :   2017-2018
************************************************
*/
class Solution
{

private:
    void dfs(vector<vector<int>> &graph, int source, int destination, vector<vector<int>> &result, vector<int> &currentPath, vector<bool> &visited)
    {
        if (source == destination)
        {
            currentPath.push_back(source);
            result.push_back(currentPath);

            currentPath.pop_back();
            return;
        }

        visited[source] = true;
        currentPath.push_back(source);

        for (int i = 0; i < graph[source].size(); i++)
            if (!visited[graph[source][i]])
                dfs(graph, graph[source][i], destination, result, currentPath, visited);

        currentPath.pop_back();
        visited[source] = false;
        return;
    }

public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>> &graph)
    {
        vector<vector<int>> result;
        vector<int> currentPath;
        vector<bool> visited(graph.size(), false);
        dfs(graph, 0, graph.size() - 1, result, currentPath, visited);
        return result;
    }
};

File: codes/leetcode/AllPathsFromSourcetoTarget.py
# ************************************************
# username  :   smmehrab
# fullname  :   s.m.mehrabul islam
# email     :   smmehrabul-2017614964@cs.du.ac.bd
# institute :   university of dhaka, bangladesh
# reg       :   2017614964
# ************************************************

class Solution:

    def _dfs(self, node, path):
        if node == self.n - 1:
            self.paths.append(path)
        else:
            for neighbor in self.graph[node]: 
                self._dfs(neighbor, path + [neighbor])

    def allPathsSourceTarget(self, graph: list[list[int]]) -> list[list[int]]:
        self.n = len(graph)
        self.graph = graph
        self.paths = []
        self._dfs(0, [0])
        return self.paths